Sources/MongoSwift/MongoClient.swift

```diff
@@ -270,14 +278,17 @@ public class MongoClient {
      *
      * - Throws:
      *   - `UserError.invalidArgumentError` if the options passed are an invalid combination.
+     *   - `UserError.logicError` if the provided session is inactive.
      *   - `EncodingError` if an error is encountered while encoding the options to BSON.
      */
-    public func listDatabases(options: ListDatabasesOptions? = nil) throws -> MongoCursor<Document> {
-        let opts = try self.encoder.encode(options)
-        guard let cursor = mongoc_client_find_databases_with_opts(self._client, opts?.data) else {
+    public func listDatabases(options: ListDatabasesOptions? = nil,
+                              session: ClientSession? = nil) throws -> MongoCursor<Document> {
+        var opts = try self.encoder.encode(options) ?? Document()
```

Comment:
This (and all the others like it) construct a new `Document` every time, regardless of whether any options or a session is passed in. I think `bson_t` does some nifty things where this is a stack allocation if the doc is small enough, so I don't think this is a big performance issue.

Reply:
I know we kind of duked this out on the design already, but something makes me uncomfortable about this approach. Could we instead make all of the options conform to some protocol that provided a `withSession(ClientSession?)`, so you would only add it in the event that the options did exist (and therefore only allocate once)? Something like:
```
guard let cursor = mongoc_client_find_databases_with_opts(self._client, opts?.withSession(session).data) else {
```

Originally I was thinking of `SessionProvider` storing the session with the options, but this would be like an "appending copy" of the options.